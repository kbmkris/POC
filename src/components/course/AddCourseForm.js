import React from "react";
import styles from "../../css/index.css";
import { PropTypes } from "prop-types";

/*
  id refers to actual field name.
  name refers to object it belongs to
  e.g. topic is an array of object with topicId & topicName as its properties
  so name will be mentioned as topic and id will be topicName
/*/
const AddCourseForm = ({
  isLoading,
  courseName,
  title,
  handleOnChange,
  handleOnTitleChange,
  handleOnTopicChange,
  handleOnClickAddTitle,
  handleOnClickRemoveTitle,
  handleOnClickAddTopic,
  handleOnClickRemoveTopic,
  handleOnClickSubmitCourse
}) =>  (
  <div className="main-div">
  <div className="addcourse-div">
    <h3> Add Course Page </h3>
    <form>
      <label
        className="label"
        htmlFor="courseName" >
        Course Name :
      </label>
      <input
        type="text"
        className="input"
        name="courseName"
        value={courseName}
        onChange={handleOnChange} />
      <br />
      {title.map((title,tidx) => (
        <div key={`d${tidx}`}>
          <div>
            <label
              className="label"
              htmlFor="courseTitle" >
              Course Title {tidx + 1}:
            </label>
            <div>
              <input
                key={`title${tidx}`}
                type="text"
                className="input"
                name="courseTitle"
                value={title.courseTitle}
                onChange={(e) => handleOnTitleChange(tidx,e)} />
              <span>
                <input
                  key={`b${tidx}1`}
                  type="button"
                  className="small-button"
                  name="addTitleButton"
                  value=" + "
                  onClick={(e) => handleOnClickAddTitle(tidx,e)} />
              </span>
              <span>
                <input
                  key={`b${tidx}2`}
                  type="button"
                  className="small-button"
                  name="removeTitleButton"
                  value=" - "
                  onClick={(e) => handleOnClickRemoveTitle(tidx,e)} />
              </span>
              </div>
          </div>
          <div>
            <label
              className="label"
              htmlFor="link" >
              Course Link {tidx + 1}:
            </label>
              <input
                key={`link${tidx}`}
                type="text"
                className="input"
                name="link"
                value={title.link}
                onChange={(e) => handleOnTitleChange(tidx,e)} />
          </div>
          {title.topic.map((topic, ttidx) => (
            <div  key={`d${tidx}t${ttidx}`}>
              <label
                className="label"
                htmlFor="topic" >
                {`     Topic Name ${tidx + 1} - ${ttidx + 1}:`}
              </label>
              <div>
                <input
                  key={`topic${tidx}t${ttidx}`}
                  type="text"
                  className="input"
                  name="topicName"
                  value={topic.topicName}
                  onChange={(e) => handleOnTopicChange(tidx,ttidx,e)} />
                <span>
                  <input
                    key={`tbutton${tidx}1`}
                    type="button"
                    className="small-button"
                    name="addTopicButton"
                    value={" + "}
                    onClick={(e) => handleOnClickAddTopic(tidx,ttidx,e)} />
                  <input
                    key={`tbutton${tidx}2`}
                    type="button"
                    className="small-button"
                    name="removeTopicButton"
                    value={" - "}
                    onClick={(e) => handleOnClickRemoveTopic(tidx,ttidx,e)} />
                </span>
              </div>
            </div>
          ))}
        </div>
      ))}
      <div className="button-div">
        <input
          type="button"
          className="button"
          name="addCourseButton"
          value={isLoading ? "Adding Course..." : "Add Course"}
          onClick={handleOnClickSubmitCourse} />
        <br />
      </div>
    </form>
  </div>
  </div>
);

AddCourseForm.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  courseName : PropTypes.string.isRequired,
  title : PropTypes.array.isRequired,
  message : PropTypes.string.isRequired,
  handleOnChange : PropTypes.func.isRequired,
  handleOnTitleChange : PropTypes.func.isRequired,
  handleOnTopicChange : PropTypes.func.isRequired,
  handleOnClickAddTitle : PropTypes.func.isRequired,
  handleOnClickRemoveTitle: PropTypes.func.isRequired,
  handleOnClickAddTopic : PropTypes.func.isRequired,
  handleOnClickRemoveTopic : PropTypes.func.isRequired,
  handleOnClickSubmitCourse : PropTypes.func.isRequired
};

AddCourseForm.defaultProps = {
  isLoading: false,
  courseName: "",
  title: [{ topic: [] }],
  message: ""
};

export default AddCourseForm;
