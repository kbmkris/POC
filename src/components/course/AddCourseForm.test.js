import React from "react";
import AddCourseForm from "./AddCourseForm";
import renderer from "react-test-renderer";
import { mount, shallow } from "enzyme";

describe("AddCourseForm Test", () => {
  it("Snapshot - component should render", () => {
    const handleOnChange = jest.fn();
    const handleOnTitleChange = jest.fn();
    const handleOnTopicChange = jest.fn();
    const handleOnClickAddTitle = jest.fn();
    const handleOnClickRemoveTitle = jest.fn();
    const handleOnClickAddTopic = jest.fn();
    const handleOnClickRemoveTopic = jest.fn();
    const handleOnClickSubmitCourse = jest.fn();

    const component = renderer
      .create(
        <AddCourseForm
          isLoading={false}
          courseName= {""}
          title= {[{ topic: []}]}
          handleOnChange={handleOnChange}
          handleOnTitleChange={handleOnTitleChange}
          handleOnTopicChange={handleOnTopicChange}
          handleOnClickAddTitle={handleOnClickAddTitle}
          handleOnClickRemoveTitle={handleOnClickRemoveTitle}
          handleOnClickAddTopic={handleOnClickAddTopic}
          handleOnClickRemoveTopic={handleOnClickRemoveTopic}
          handleOnClickSubmitCourse={handleOnClickSubmitCourse}
        />,
      );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe ("Dom validation using enzyme", () => {
  function setup (isLoading) {
    const props = {
      isLoading: isLoading,
      courseName: "HTML",
      title: [
        {
          courseTitleId: 1,
          courseTitle: "Fundamental",
          link: "https://stackoverflow.com/html",
          courseId: 1,
          topic: [
            {
              topicId: 1,
              topicName: "Basic tag"
            },
            {
              topicId: 2,
              topicName: "Form"
            }
          ]
        },
        {
          courseTitleId: 2,
          courseTitle: "Advanced Feature",
          link: "https://stackoverflow.com/advanced",
          courseId: 1,
          topic: [
            {
              topicId: 3,
              topicName: "Grid"
            }
          ]
        }
      ],
      handleOnChange: jest.fn(),
      handleOnTitleChange: jest.fn(),
      handleOnTopicChange: jest.fn(),
      handleOnClickAddTitle: jest.fn(),
      handleOnClickRemoveTitle: jest.fn(),
      handleOnClickAddTopic: jest.fn(),
      handleOnClickRemoveTopic: jest.fn(),
      handleOnClickSubmitCourse: jest.fn(),
    };
    return shallow(<AddCourseForm {...props} />);
  }
  it("check for all elements inside AddCourseForm", () => {
    const wrapper = setup(false);
    expect(wrapper.find("h3").text()).toEqual(" Add Course Page ");
    expect(wrapper.find({name: "courseName"}).length).toBe(1);
    expect(wrapper.find({name: "courseTitle"}).length).toBe(2); // equal to no of titles in props
    expect(wrapper.find({name: "addTitleButton"}).length).toBe(2); //equal to no of titles in props
    expect(wrapper.find({name: "removeTitleButton"}).length).toBe(2); //equal to no of titles in props
    expect(wrapper.find({name: "topicName"}).length).toBe(3); //equal to no of topics in props
    expect(wrapper.find({name: "addTopicButton"}).length).toBe(3); //equal to no of topics in props
    expect(wrapper.find({name: "removeTopicButton"}).length).toBe(3); //equal to no of topics in props
    expect(wrapper.find({name: "addCourseButton"}).length).toBe(1);//one add course button
    console.log(wrapper.find({name: "courseName"}).debug());
  });

  it("check whether Add course button value changes on click", () => {
    const wrapper = setup(true);
    const addCourseButton = wrapper.find({name: "addCourseButton"});
    addCourseButton.simulate("click");
    expect(addCourseButton.prop("value")).toEqual("Adding Course...");
  });
});
