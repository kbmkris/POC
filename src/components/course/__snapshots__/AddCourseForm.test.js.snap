// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCourseForm Test Snapshot - component should render 1`] = `
<div
  className="main-div"
>
  <div
    className="addcourse-div"
  >
    <h3>
       Add Course Page 
    </h3>
    <form>
      <label
        className="label"
        htmlFor="courseName"
      >
        Course Name :
      </label>
      <input
        className="input"
        name="courseName"
        onChange={[MockFunction]}
        type="text"
        value=""
      />
      <br />
      <div>
        <div>
          <label
            className="label"
            htmlFor="courseTitle"
          >
            Course Title 
            1
            :
          </label>
          <div>
            <input
              className="input"
              name="courseTitle"
              onChange={[Function]}
              type="text"
            />
            <span>
              <input
                className="small-button"
                name="addTitleButton"
                onClick={[Function]}
                type="button"
                value=" + "
              />
            </span>
            <span>
              <input
                className="small-button"
                name="removeTitleButton"
                onClick={[Function]}
                type="button"
                value=" - "
              />
            </span>
          </div>
        </div>
        <div>
          <label
            className="label"
            htmlFor="link"
          >
            Course Link 
            1
            :
          </label>
          <input
            className="input"
            name="link"
            onChange={[Function]}
            type="text"
          />
        </div>
      </div>
      <div
        className="button-div"
      >
        <input
          className="button"
          name="addCourseButton"
          onClick={[MockFunction]}
          type="button"
          value="Add Course"
        />
        <br />
      </div>
    </form>
  </div>
</div>
`;
